File: src/readers/gbfs/schemaV1/systemAlerts.ts
/**
 * # GBFS System Alerts Schema V1.1 OR GBFS System Alerts Schema V1.0
 * Describes ad-hoc changes to the system.
 *
 * ## Links
 * - [GBFS Specification V1.1](https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#system_alertsjson)
 * - [GBFS Specification V1.0](https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#system_alertsjson)
 */
export type GBFSSystemAlertsV1 = GBFSSystemAlertsV11 | GBFSSystemAlertsV10;

/**
 * # GBFS System Alerts Schema V1.1
 * Describes ad-hoc changes to the system.
 *
 * ## Links
 * - [GBFS Specification V1.1](https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#system_alertsjson)
 */
export const gbfsSystemAlertsSchemaV11 = {
  $schema: 'http://json-schema.org/draft-07/schema',
  $id: 'https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#system_alertsjson',
  description: 'Describes ad-hoc changes to the system.',
  type: 'object',
  properties: {
    last_updated: {
      description: 'Last time the data in the feed was updated in POSIX time.',
      type: 'integer',
      minimum: 1450155600,
    },
    ttl: {
      description:
        'Number of seconds before the data in the feed will be updated again (0 if the data should always be refreshed).',
      type: 'integer',
      minimum: 0,
    },
    version: {
      description:
        'GBFS version number to which the feed conforms, according to the versioning framework (added in v1.1).',
      type: 'string',
      const: '1.1',
    },
    data: {
      description: 'Array that contains ad-hoc alerts for the system.',
      type: 'object',
      properties: {
        alerts: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              alert_id: { description: 'Identifier for this alert.', type: 'string' },
              type: {
                description: 'Type of alert.',
                type: 'string',
                enum: ['SYSTEM_CLOSURE', 'STATION_CLOSURE', 'STATION_MOVE', 'OTHER'],
              },
              times: {
                description: 'Array of objects indicating when the alert is in effect.',
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    start: {
                      description: 'Start time of the alert.',
                      type: 'number',
                      minimum: 1450155600,
                    },
                    end: {
                      description: 'End time of the alert.',
                      type: 'number',
                      minimum: 1450155600,
                    },
                  },
                },
                additionalItems: false,
                required: ['start'],
              },
              station_ids: {
                description: 'Array of identifiers of the stations for which this alert applies.',
                type: 'array',
                items: { type: 'string' },
              },
              region_ids: {
                description: 'Array of identifiers of the regions for which this alert applies.',
                type: 'array',
                items: { type: 'string' },
              },
              url: {
                description: 'URL where the customer can learn more information about this alert.',
                type: 'string',
                format: 'uri',
              },
              summary: {
                description: 'A short summary of this alert to be displayed to the customer.',
                type: 'string',
              },
              description: {
                description: 'Detailed description of the alert.',
                type: 'string',
              },
              last_updated: {
                description: 'Indicates the last time the info for the alert was updated.',
                type: 'number',
                minimum: 1450155600,
              },
            },
            required: ['alert_id', 'type', 'summary'],
          },
        },
      },
      required: ['alerts'],
    },
  },
  required: ['last_updated', 'ttl', 'version', 'data'],
};

/**
 * GBFS System Alerts Schema V1.1 Interface
 */
export interface GBFSSystemAlertsV11 {
  /** Last time the data in the feed was updated in POSIX time. */
  last_updated: number;

  /** Number of seconds before the data in the feed will be updated again. */
  ttl: number;

  /** GBFS version number (1.1). */
  version: '1.1';

  /** Data containing ad-hoc alerts for the system. */
  data: {
    alerts: Array<{
      alert_id: string;
      type: 'SYSTEM_CLOSURE' | 'STATION_CLOSURE' | 'STATION_MOVE' | 'OTHER';
      times?: Array<{
        start: number;
        end?: number;
      }>;
      station_ids?: string[];
      region_ids?: string[];
      url?: string;
      summary: string;
      description?: string;
      last_updated?: number;
    }>;
  };
}

/**
 * # GBFS System Alerts Schema V1.0
 * Describes ad-hoc changes to the system.
 *
 * ## Links
 * - [GBFS Specification V1.0](https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#system_alertsjson)
 */
export const gbfsSystemAlertsSchemaV10 = {
  $schema: 'http://json-schema.org/draft-07/schema',
  $id: 'https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#system_alertsjson',
  description: 'Describes ad-hoc changes to the system.',
  type: 'object',
  properties: {
    last_updated: {
      description: 'Last time the data in the feed was updated in POSIX time.',
      type: 'integer',
      minimum: 0,
      maximum: 1924988399,
    },
    ttl: {
      description:
        'Number of seconds before the data in the feed will be updated again (0 if the data should always be refreshed).',
      type: 'integer',
      minimum: 0,
    },
    data: {
      description: 'Array that contains ad-hoc alerts for the system.',
      type: 'object',
      properties: {
        alerts: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              alert_id: { description: 'Identifier for this alert.', type: 'string' },
              type: {
                description: 'Type of alert.',
                type: 'string',
                enum: ['SYSTEM_CLOSURE', 'STATION_CLOSURE', 'STATION_MOVE', 'OTHER'],
              },
              times: {
                description: 'Array of objects indicating when the alert is in effect.',
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    start: { type: 'number', minimum: 0 },
                    end: { type: 'number', minimum: 0 },
                  },
                },
                additionalItems: false,
                required: ['start'],
              },
              station_ids: {
                description: 'Array of identifiers of the stations for which this alert applies.',
                type: 'array',
                items: { type: 'string' },
              },
              region_ids: {
                description: 'Array of identifiers of the regions for which this alert applies.',
                type: 'array',
                items: { type: 'string' },
              },
              url: {
                description: 'URL where the customer can learn more information about this alert.',
                type: 'string',
              },
              summary: {
                description: 'A short summary of this alert to be displayed to the customer.',
                type: 'string',
              },
              description: {
                description: 'Detailed description of the alert.',
                type: 'string',
              },
              last_updated: {
                description: 'Indicates the last time the info for the alert was updated.',
                type: 'integer',
                minimum: 0,
                maximum: 1924988399,
              },
            },
            required: ['alert_id', 'type', 'summary'],
          },
        },
      },
      required: ['alerts'],
    },
  },
  required: ['last_updated', 'ttl', 'data'],
};

/**
 * GBFS System Alerts Schema V1.0 Interface
 */
export interface GBFSSystemAlertsV10 {
  /** Last time the data in the feed was updated in POSIX time. */
  last_updated: number;

  /** Number of seconds before the data in the feed will be updated again. */
  ttl: number;

  /** Data containing ad-hoc alerts for the system. */
  data: {
    alerts: Array<{
      alert_id: string;
      type: 'SYSTEM_CLOSURE' | 'STATION_CLOSURE' | 'STATION_MOVE' | 'OTHER';
      times?: Array<{
        start: number;
        end?: number;
      }>;
      station_ids?: string[];
      region_ids?: string[];
      url?: string;
      summary: string;
      description?: string;
      last_updated?: number;
    }>;
  };
}

File: rust/readers/gbfs/schema_v3/system_alerts.rs
use crate::readers::GBFSName;
use alloc::{string::String, vec::Vec};
use serde::{Deserialize, Serialize};

/// # GBFS System Alerts Schema V3.1-RC & V3.0
/// Describes ad-hoc changes to the system.
///
/// ## Links
/// - [GBFS Specification V3.1-RC](https://github.com/MobilityData/gbfs/blob/v3.1-RC/gbfs.md#system_alertsjson)
/// - [GBFS Specification V3.0](https://github.com/MobilityData/gbfs/blob/v3.0/gbfs.md#system_alertsjson)
pub type GBFSSystemAlertsV3 = GBFSSystemAlertsV30;

/// GBFS System Alerts Alert Type
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub enum GBFSSystemAlertsAlertTypeV30 {
    /// System Closure
    #[serde(rename = "system_closure")]
    SystemClosure,
    /// Station Closure
    #[serde(rename = "station_closure")]
    StationClosure,
    /// Station Move
    #[serde(rename = "station_move")]
    StationMove,
    /// Other
    #[default]
    #[serde(rename = "other")]
    Other,
}

/// GBFS System Alerts Alert Times
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSSystemAlertsAlertTimesV30 {
    /// Start time in POSIX time
    pub start: u64,
    /// End time in POSIX time
    pub end: Option<u64>,
}

/// GBFS System Alerts Alert
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSSystemAlertsAlertV30 {
    /// Identifier for this alert.
    pub alert_id: String,
    /// Type of alert.
    /// Possible values: 'system_closure', 'station_closure', 'station_move', 'other'.
    pub r#type: GBFSSystemAlertsAlertTypeV30,
    /// Array of objects indicating when the alert is in effect.
    pub times: Option<Vec<GBFSSystemAlertsAlertTimesV30>>,
    /// Array of identifiers of the stations for which this alert applies.
    pub station_ids: Option<Vec<String>>,
    /// Array of identifiers of the regions for which this alert applies.
    pub region_ids: Option<Vec<String>>,
    /// URL where customers can learn more information about this alert.
    pub url: Option<Vec<GBFSName>>,
    /// Short summary of this alert to be displayed to the customer.
    pub summary: Vec<GBFSName>,
    /// Detailed description of the alert.
    pub description: Option<Vec<GBFSName>>,
    /// Indicates the last time the info for the alert was updated in RFC3339 format.
    /// **Format**: date-time
    pub last_updated: Option<String>,
}

/// GBFS System Alerts Alerts
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSSystemAlertsAlertsV30 {
    /// Data containing ad-hoc alerts for the system.
    pub alerts: Vec<GBFSSystemAlertsAlertV30>,
}

/// # GBFS System Alerts Schema V3.0
/// Describes ad-hoc changes to the system.
///
/// ## Links
/// - [GBFS Specification](https://github.com/MobilityData/gbfs/blob/v3.0/gbfs.md#system_alertsjson)
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSSystemAlertsV30 {
    /// Last time the data in the feed was updated in RFC3339 format.
    /// **Format**: date-time
    pub last_updated: String,
    /// Number of seconds before the data in the feed will be updated again (0 if the data should always be refreshed).
    /// **Minimum**: 0
    pub ttl: u64,
    /// GBFS version number to which the feed conforms.
    /// **Const**: '3.0'
    pub version: String,
    /// Data object containing system alerts.
    pub data: GBFSSystemAlertsAlertsV30,
}
